### PR TITLE
Retry GHES archive generation process in case of any error

### LINKED ISSUE #611: Retry GHES archive generation
We've been seeing a bunch of integration test failures when generating GHES archives. Usually the error is something like the connection was closed, or other networking sounding errors.

We could probably do a retry on archive generation if it fails for any reason.

### PR DESCRIPTION
**Pull request by ArinGhazarian**
- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created)
- [x] New package licenses are added to `ThirdPartyNotices.txt` (if applicable)

Closes #611 

This PR adds a more robust retry logic for GHES archive generation process by retrying the entire process in case of any failure (Network timeout, Archive generation failure, etc.). The following changes were made:
1. All changes made as part of #801 were reverted because we were trying to retry a non transient error for a `failed` archive generation result, so it would just return the same result over and over again!
2. Some refactors were made to extract out the archive generation process into its own method to be able to easily wrap it in  a retry policy. 
3. Simplified `Retry` method in `RetryPolicy` and some minor refactors. 
4. Added a unit test to cover various retries for GHES archive generation.

### CODE REVIEW

File: src/gei/Handlers/MigrateRepoCommandHandler.cs [hunk 1]
```diff
@@ -188,56 +199,67 @@ private string ExtractGhesBaseUrl(string ghesApiUrl)
...
 
             _log.LogInformation($"Downloading archive from {metadataArchiveUrl}");
-            await _httpDownloadService.DownloadToFile(metadataArchiveUrl, metadataArchiveFilePath);
+            await _httpDownloadService.DownloadToFile(metadataArchiveUrl, metadataArchiveDownloadFilePath);
 
 #pragma warning disable IDE0063
-            await using (var gitArchiveContent = _fileSystemProvider.OpenRead(gitArchiveFilePath))
-            await using (var metadataArchiveContent = _fileSystemProvider.OpenRead(metadataArchiveFilePath))
+            await using (var gitArchiveContent = _fileSystemProvider.OpenRead(gitArchiveDownloadFilePath))
+            await using (var metadataArchiveContent = _fileSystemProvider.OpenRead(metadataArchiveDownloadFilePath))
```
[the remark this answers is not in the document]
Comment (ArinGhazarian) [reply src/gei/Handlers/MigrateRepoCommandHandler.cs:225]: Yes the newer using syntax is def. more sleek but there there are situations where you want to immediately dispose the resource as opposed to when the method execution is done. Here is we don't immediately dispose, the file handles are going to be in use so the files cannot be deleted.

File: src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs [hunk 2]
```diff
@@ -849,6 +853,96 @@ await FluentActions
...
+            var metadataArchiveId = 2;
+            var gitArchiveUrl = $"https://example.com/{gitArchiveId}";
+            var metadataArchiveUrl = $"https://example.com/{metadataArchiveId}";
+            var authenticatedGitArchiveUrl = new Uri($"https://example.com/{gitArchiveId}/authenticated");
+            var authenticatedMetadataArchiveUrl = new Uri($"https://example.com/{metadataArchiveId}/authenticated");
+            var gitArchiveFilePath = "path/to/git_archive";
+            var metadataArchiveFilePath = "path/to/metadata_archive";
+
+            _mockSourceGithubApi.Setup(x => x.GetEnterpriseServerVersion()).ReturnsAsync("3.7.1");
+            _mockTargetGithubApi.Setup(x => x.GetOrganizationId(TARGET_ORG).Result).Returns(githubOrgId);
```
[the remark this answers is not in the document]
Comment (ArinGhazarian) [reply src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs:875]: see [this](https://github.com/Moq/moq4/wiki/Quickstart#async-methods) for `ReturnAsync()` vs `.Result).Returns`

[the remark this answers is not in the document]
Comment (dylan-smith) [reply src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs:875]: yeah, I'm just suggesting in the new code/test here we use ReturnsAsync

[the remark this answers is not in the document]
Comment (dylan-smith) [reply src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs:875]: huh interesting, sounds like Moq is suggesting the .Result way. so be it.

**Review by dylan-smith — CHANGES_REQUESTED** [review]
this all looks good to me with one change. When you merge with the other change to RetryPolicy already in `main` lets change it to not return the PolicyResult and directly return the value or throw an exception. At some point in the future we should probably do the same to `RetryOnResult` too

**Review by dylan-smith — APPROVED** [review]

**Pull request merged**